### ISSUE TITLE
Avoiding derived attributes

### ISSUE THREAD

**Issue by gadgetmg** [acrmp/foodcritic#386]
This is an issue I've come across from time to time with some fairly high profile cookbooks. I think it should be considered an anti-pattern to derive attributes from other attributes.

For example, the [postgresql cookbook](https://github.com/hw-cookbooks/postgresql) is an example as to why this is troublesome. Since it uses the ```version``` attribute in other attributes, setting the default ```version``` attribute in a wrapper cookbook doesn't change derivative attributes because Chef evaluates them before the wrapper cookbook's attributes are loaded.

A couple of blogs have talked about the issue and discussed better ways to handle attributes:
* https://christinemdraper.wordpress.com/2014/10/06/avoiding-the-possible-pitfalls-of-derived-attributes/
* https://coderanger.net/derived-attributes/

It would probably be good to have a foodcritic rule warning of this in order to make the lives of those writing wrapper cookbooks better.

**Comment by jaymzh** [comment]
I agree.... to an extent. :)

It's not the derived attributes is bad, it's about how you use them. If you do them in the attributes file it's very bad. If you do the deriving in a recipe, at least people can set them in an earlier cookbook. And my preference is you only "derive" at runtime, when you're going to use the derived value. There is, of course, some reasons for using all of these.

I bring this up, because depending on your model this could land people in a world of false-positives... so while I think this is a fantastic idea... we need to be careful.

**Comment by gadgetmg** [comment]
Agreed. It should definitely be avoided in the attributes file because of the order Chef calculates them. In a recipe it's not a problem.

**Comment by jaymzh** [comment]
Funny story - the whole reason that the Chef 11 attributes changed as they did was to support derived attributes in attributes files because people wanted to write cookbooks that used them, where people overrided them in roles. Sadpanda.

**Comment by lamont-granquist** [comment]
Not quite correct history, it was because people wanted to write library cookbooks with attributes files which override attributes directly set in recipe code in other cookbooks.  Wasn't to support derived attributes.

These days we mostly push people to write library code using resources and not attributes or recipes which renders the whole question moot.  Even back then, though, solving the problem of having badly written library cookbooks with hardcoded attribute values in recipe code was probably something we never should have 'solved' -- but we had users screaming at us that our wrapper cookbook philosophy was horrible and broken and they couldn't manage their infrastructure the way we telling them to (even though we didn't come up with that idea and it was just the community that was running that direction as fast as they could...)

I'm leaning towards derived attributes being bad, though.  You can always not use an attribute and derive the value directly in recipe code and shove it into a plain-old-ruby-variable and solve this problem in a way that gives the API that people expect.
